x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/whitelist/InternalSqlScriptUtils.java

```diff
@@ -548,6 +548,6 @@ public static Double stZ(Object g) {
     public static Object cast(Object value, String typeName) {
         // we call asDateTime here to make sure we handle JodaCompatibleZonedDateTime properly,
         // since casting works for ZonedDateTime objects only
-        return DataTypeConversion.convert(asDateTime(value, true), DataType.fromTypeName(typeName));
+        return DataTypeConversion.convert(asDateTime(value, true), DataType.fromSqlOrEsType(typeName));
```

Comment:
I wonder if the `DateType.fromTypeName` method shouldn't be investigated as I see it's still being used inside the code in various places related to scripting.